### PR TITLE
Add gradient to SignedDistance::Ellipsoid

### PR DESCRIPTION
**Pull request by bergbauer**
FYI @nmuch

### CODE REVIEW

File: source/base/function_signed_distance.cc [hunk 1]
```diff
@@ -166,6 +166,42 @@ namespace Functions
...
+                             const unsigned int component) const
+    {
+      AssertIndexRange(component, this->n_components);
+      (void)component;
+
+      const Point<dim> point_in_centered_cosy =
+        Point<dim>(compute_closest_point_ellipse(point) - center);
+
+      Tensor<1, dim> grad =
+        compute_analyical_normal_vector_on_ellipse(point_in_centered_cosy);
```
Comment (bangerth) [inline source/base/function_signed_distance.cc:181]:
```suggestion
      const Tensor<1, dim> grad =
        compute_analyical_normal_vector_on_ellipse(point_in_centered_cosy);
```

File: include/deal.II/base/function_signed_distance.h [hunk 2]
```diff
@@ -161,6 +161,22 @@ namespace Functions
...
+      /**
+       * Calculates the gradient of the signed distance function via the normal
+       * of the closest point on the ellipsoid.
+       */
+      Tensor<1, dim>
+      gradient(const Point<dim> &,
+               const unsigned int component = 0) const override;
+
+      /**
+       * Estimates the hessian of the signed distance function via the original
```
Comment (bangerth) [inline include/deal.II/base/function_signed_distance.h:173]:
```suggestion
       * Estimates the Hessian of the signed distance function via the original
```

File: source/base/function_signed_distance.cc [hunk 3]
```diff
@@ -166,6 +166,42 @@ namespace Functions
...
 
+    template <int dim>
+    Tensor<1, dim>
+    Ellipsoid<dim>::gradient(const Point<dim> & point,
+                             const unsigned int component) const
+    {
+      AssertIndexRange(component, this->n_components);
+      (void)component;
+
+      const Point<dim> point_in_centered_cosy =
```
Comment (bangerth) [inline source/base/function_signed_distance.cc:177]: What does the term `cosy` refer to here?

File: source/base/function_signed_distance.cc [hunk 4]
```diff
@@ -166,6 +166,42 @@ namespace Functions
...
+        Point<dim>(compute_closest_point_ellipse(point) - center);
+
+      Tensor<1, dim> grad =
+        compute_analyical_normal_vector_on_ellipse(point_in_centered_cosy);
+      return grad / grad.norm();
```
Comment (bangerth) [inline source/base/function_signed_distance.cc:182]: I don't understand this. Presumably, the function is of the form `(x/a)^2+(y/b)^2`. Then its gradient must vanish at the center of the ellipse, but you're making sure that the magnitude of the gradient is always one...

File: source/base/function_signed_distance.cc [hunk 5]
```diff
@@ -166,6 +166,42 @@ namespace Functions
...
+    Ellipsoid<dim>::hessian(const Point<dim> &,
+                            const unsigned int component) const
+    {
+      AssertIndexRange(component, this->n_components);
+      (void)component;
+
+      SymmetricTensor<2, dim> hess = unit_symmetric_tensor<dim>();
+
+      hess[0][0] = 2 / radii[0] * radii[0];
+      hess[1][1] = 2 / radii[1] * radii[1];
```
Comment (bangerth) [inline source/base/function_signed_distance.cc:198]: This only works for 2d functions. But it shouldn't be difficult to write this via a `for` loop in a dimension independent way.

Separately, though, the computation is not correct. You want this:
```suggestion
      hess[0][0] = 2 / (radii[0] * radii[0]);
      hess[1][1] = 2 / (radii[1] * radii[1]);
```

File: source/base/function_signed_distance.cc [hunk 6]
```diff
@@ -245,7 +278,59 @@ namespace Functions
       while (std::abs(delta_t) > tolerance && iter < max_iter);
       AssertIndexRange(iter, max_iter);
 
-      const double distance = std::hypot(x - px, y - py);
+      AssertIsFinite(x);
+      AssertIsFinite(y);
+
+      return Point<dim>(sign_px * x + center[0], sign_py * y + center[1]);
```
Comment (bangerth) [inline source/base/function_signed_distance.cc:284]: Maybe easier to read:
```suggestion
      return (center + Point<dim>(sign_px * x, sign_py * y));
```

**Review by bangerth — CHANGES_REQUESTED** [review]
I've got a few comments. I think it would be useful if there were a test, though, that checks the correctness of the implementation.

File: source/base/function_signed_distance.cc (hunk 3 above)
Comment (bergbauer) [reply source/base/function_signed_distance.cc]: For me cosy is an abbreviation for coordinate system but it is probably not worth the possible confusion. I changed it and spelled it out.

File: source/base/function_signed_distance.cc (hunk 4 above)
Comment (bergbauer) [reply source/base/function_signed_distance.cc:182]: image
The signed distance function looks like this and is not `(x/a)^2+(y/b)^2 - 1 = 0`. I think the implemented gradient is correct.

File: source/base/function_signed_distance.cc (hunk 5 above)
Comment (bergbauer) [reply source/base/function_signed_distance.cc]: Playing around with my code today showed that this is not robust enough to estimate function bounds correctly for all cases. I suggest to use the hessian of a sphere with radius `(a + b) / 2` (see last commit).

**Comment by bergbauer** [comment]
Thanks for the review @bangerth ! What would you like to see as a test? The gradient computation on the semi-minor and semi-major axis?

File: source/base/function_signed_distance.cc (hunk 5 above)
Comment (bergbauer) [reply source/base/function_signed_distance.cc]: If anybody has a better idea how to compute a better approximation or even the analytic Hessian, suggestions are welcome!

**Comment by peterrum** [comment]
@bergbauer  How should we proceed here? Is this PR still relevant given #13616?

**Comment by bergbauer** [comment]
> How should we proceed here? Is this PR still relevant given https://github.com/dealii/dealii/pull/13616?

Not necessarily, you can use a discrete level set vector to compute the gradient and the hessian cellwise with the other PR. The analytic gradient of the signed distance function might be nice to have, the hessian is only an approximation and does not really fit in the library well. I would not mind closing this.

**Comment by peterrum** [comment]
> The analytic gradient of the signed distance function might be nice to have, the hessian is only an approximation and does not really fit in the library well. I would not mind closing this.

I would keep gradient if it is analytical. I would remote the approximation of the Hessian.

**Comment by kronbichler** [comment]
I agree, let us keep the analytic gradient.

**Comment by bergbauer** [comment]
A potential test could be to check the analytic gradient against the gradient of a finite element with discrete level set values.

**Review by peterrum — APPROVED** [review]
> A potential test could be to check the analytic gradient against the gradient of a finite element with discrete level set values.

Sounds good! Ready after you added the test!

**Comment by bangerth** [comment]
Alternatively, if you check by hand that the implementation you have is "reasonable", then just outputting the currently computed value into the output file is acceptable. We have many tests for which we don't know the exact solution, and in these cases tests strictly speaking do not check that the answer is correct, but only that the answer does not change with subsequent commits.

**Comment by peterrum** [comment]
@bergbauer Could you finalize this PR?

**Comment by bergbauer** [comment]
@peterrum @bangerth I added the output of the gradient to the existing test. If we additionally want the correctness check just let me know.

**Review by kronbichler — APPROVED** [review]

**Review by peterrum — APPROVED** [review]

**Pull request merged**